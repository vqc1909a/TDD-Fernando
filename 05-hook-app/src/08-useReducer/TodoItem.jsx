
import PropTypes from "prop-types";
import React from "react";

export const TodoItem = React.memo(({description, id, done ,handleDeleteTodo, handleToggleTodo}) => {
  return (
    <li aria-label="li" className="list-group-item d-flex justify-content-between">
      <span aria-label="span" style={{cursor: 'pointer'}} className={`align-self-center flex-fill ${done ? 'text-decoration-line-through': ''}`} onClick={() => handleToggleTodo(id)}>{description}</span>
      <button aria-label="button" className="btn btn-danger" onClick={() => handleDeleteTodo(id)}>Borrar</button>
    </li>
    )
})


TodoItem.propTypes = {
  description: PropTypes.string,
  id: PropTypes.string,
  done: PropTypes.bool,
  handleDeleteTodo: PropTypes.func.isRequired,
  handleToggleTodo: PropTypes.func.isRequired,
}

