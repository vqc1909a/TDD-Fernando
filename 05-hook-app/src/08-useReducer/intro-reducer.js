const initialState = [{
    id: 1,
    todo: 'Recolectar la piedra del alma',
    donce: false,
}]

const todoReducer = (state = initialState, action = {}) => {
    if(action.type === "add"){
        return [...state, action.payload]
    }
    //The reducer always return a new state object, it mean return a new object,  new array, etc whithout matter its content in these previous cases
    return state
}


let todos = todoReducer()
const newTodo = {
    id: 2,
    todo: 'Recolectar la piedra del poder',
    done: false
}

const addTodoAction = {
    type: 'add',
    payload: newTodo
}

todos = todoReducer(todos, addTodoAction);

console.log({
    todos
})