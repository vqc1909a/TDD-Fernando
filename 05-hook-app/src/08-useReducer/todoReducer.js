
export const todoReducer = (initialState = [], action) => {
  switch (action.type) {
    case 'addTodo':
        return [
            ...initialState,
            action.payload
        ]   
    case 'deleteTodo':
        return initialState.filter(todo => todo.id !== action.payload)

    case 'toggleTodo':
        return initialState.map(todo => todo.id.toString() === action.payload.toString() ? ({...todo, done: !todo.done}) : todo)

    default:
        return initialState
  }
    
}
