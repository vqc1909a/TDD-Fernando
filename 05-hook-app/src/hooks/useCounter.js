import { useState } from "react"

export const useCounter = (initialValue = 10) => {
  const [counter, setCounter] = useState(initialValue); 

  const handleAddCounter = (value = 1) => {
    setCounter(counter + value);
  }

  const handleResetCounter = (value = initialValue) => {
    setCounter(value);
  }
  const handleSubtractCounter = (value = 1) => {
    if(counter === 0) return;
    setCounter(counter - value);
  }
  return {
    counter,
    handleAddCounter,
    handleResetCounter,
    handleSubtractCounter
  }
}
