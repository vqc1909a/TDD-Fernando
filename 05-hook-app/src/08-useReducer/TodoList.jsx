
import PropTypes from "prop-types";
import { TodoItem } from "./TodoItem";
import React from "react";

export const TodoList = React.memo(({todos, handleDeleteTodo, handleToggleTodo}) => {
  // console.log("Ejecutando todolist")
  return (
    <ul className="list-group">
        {todos.map(todo => (
            <TodoItem key={todo.id} {...todo} handleDeleteTodo={handleDeleteTodo} handleToggleTodo={handleToggleTodo}/>
        ))}
    </ul>
    )
})


TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  handleDeleteTodo: PropTypes.func.isRequired,
  handleToggleTodo: PropTypes.func.isRequired
}