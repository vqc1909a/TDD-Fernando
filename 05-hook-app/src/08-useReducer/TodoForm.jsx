
import React  from "react";
import PropTypes from "prop-types";
import { useForm } from "../hooks";
import uuid4 from "uuid4";

export const TodoForm = React.memo(({handleAddTodo}) => {
  const {description, handleInputChange, handleResetForm} = useForm({
    description: ''
  });

  const handleSubmit = (e) => {
    e.preventDefault();
    if(description.trim().length <= 0) return;

    const newTodo = {
      id: uuid4(),
      description,
      done: false
    }
    handleAddTodo(newTodo)
    handleResetForm();
  }

 
  return (
     <form onSubmit={handleSubmit}>
        <input 
            type="text" 
            placeholder="¿Qué hay que hacer?"
            className="form-control"
            name="description"
            value={description}
            onChange={handleInputChange}
        />
        <button 
            type="submit"
            className="btn btn-outline-primary mt-1"
        >
            Agregar
        </button>
    </form>
    )
})


TodoForm.propTypes = {
  handleAddTodo: PropTypes.func.isRequired
}