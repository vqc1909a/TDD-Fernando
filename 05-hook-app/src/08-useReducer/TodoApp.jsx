import {TodoList} from "./TodoList";
import { TodoForm } from "./TodoForm";
import { useTodos } from "../hooks";



export const TodoApp = () => {
    //It's recommended that we use the hook useTodos only in the father component, and then we pass the functions and the state to the children components, because we use the hook in the children components, will have two or more instances of own hook and it will trigger a leak memory
    const { handleAddTodo, handleDeleteTodo, handleToggleTodo, todos, todosCount, pendingTodosCount } = useTodos();
  
  return (
    <>
     <h1>TodoApp ({todosCount}) <small>pendientes: {pendingTodosCount}</small></h1>
     <hr />

     <div className="row">
        <div className="col-7">
            <TodoList todos={todos} handleDeleteTodo={handleDeleteTodo} handleToggleTodo={handleToggleTodo}/>
        </div>
        <div className="col-5">
            <h4>Agregar TODO</h4>
            <hr />
            <TodoForm handleAddTodo={handleAddTodo} />
        </div>
     </div>

    </>
  )
}
