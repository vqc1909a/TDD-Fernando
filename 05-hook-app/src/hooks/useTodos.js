import { useCallback, useEffect, useReducer } from "react";
import { todoReducer } from "../08-useReducer/todoReducer";

const init = () => JSON.parse(localStorage.getItem('todos')) || []

export const useTodos = () => {
    //We use the third argument of the useReducer to initialize the state of the reducer    
    const [todos, dispatch] = useReducer(todoReducer, [], init); 

    //The recommended is we have my functions that changed the state in the father component
    const handleAddTodo = useCallback((todo) => {
        dispatch({
            type: "addTodo",
            payload: todo
        })
    }, []);

    const handleDeleteTodo = (id) => {
        dispatch({
            type: "deleteTodo",
            payload: id
        })
    };

    const handleToggleTodo = (id) => {
        dispatch({
            type: "toggleTodo",
            payload: id
        })
    }

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos])

    return {
        todos,
        todosCount: todos.length,
        pendingTodosCount: todos.filter(todo => !todo.done).length,
        handleAddTodo,
        handleDeleteTodo,
        handleToggleTodo
    }
}
