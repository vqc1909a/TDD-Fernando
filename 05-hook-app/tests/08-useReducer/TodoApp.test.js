import { render } from "@testing-library/react";
import { TodoApp} from "../../src/08-useReducer/TodoApp"
import { useTodos } from "../../src/hooks/useTodos";

jest.mock("../../src/hooks/useTodos");
describe('tests on TodoApp component', () => {
    
    useTodos.mockReturnValue({
        todos: [
            {id: "1", description: "Todo #1", done: true},
            {id: "2", description: "Todo #2", done: false},
        ],
        todosCount: 2,
        pendingTodosCount: 1,
        handleAddTodo: jest.fn(),
        handleDeleteTodo: jest.fn(),
        handleToggleTodo: jest.fn(),
    })

    //Just is neccesary show the component by default because the components inside the component TodoApp will be tested already 
    test('should show the component by default', () => { 
        const { container } = render(<TodoApp />);
        expect(container).toMatchSnapshot();
    })
})