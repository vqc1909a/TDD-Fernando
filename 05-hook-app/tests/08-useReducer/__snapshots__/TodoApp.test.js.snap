// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on TodoApp component should show the component by default 1`] = `
<div>
  <h1>
    TodoApp (
    2
    ) 
    <small>
      pendientes: 
      1
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          aria-label="li"
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center flex-fill text-decoration-line-through"
            style="cursor: pointer;"
          >
            Todo #1
          </span>
          <button
            aria-label="button"
            class="btn btn-danger"
          >
            Borrar
          </button>
        </li>
        <li
          aria-label="li"
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center flex-fill "
            style="cursor: pointer;"
          >
            Todo #2
          </span>
          <button
            aria-label="button"
            class="btn btn-danger"
          >
            Borrar
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <form>
        <input
          aria-label="input"
          class="form-control"
          name="description"
          placeholder="¿Qué hay que hacer?"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
